google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java

```diff
@@ -260,6 +263,79 @@ public void downloadTo(Path path) {
     downloadTo(path, new BlobSourceOption[0]);
   }
 
+  /**
+   * Uploads the given file path to this blob using specified blob write options.
+   *
+   * @param path file to be uploaded
+   * @param options blob write options
+   * @return updated blob
+   * @throws IOException on I/O error
+   * @throws StorageException on failure
+   */
+  public Blob uploadFrom(Path path, BlobWriteOption... options) throws IOException {
+    if (Files.isDirectory(path)) {
+      throw new StorageException(0, path + ": Is a directory");
+    }
+    try (InputStream input = Files.newInputStream(path)) {
+      return uploadFrom(input, options);
+    }
+  }
+
+  /**
+   * Uploads the given content to this blob using specified blob write options.
+   *
+   * @param input content to be uploaded
+   * @param options blob write options
+   * @return updated blob
+   * @throws IOException on I/O error
+   * @throws StorageException on failure
+   */
+  public Blob uploadFrom(InputStream input, BlobWriteOption... options) throws IOException {
+    try (WriteChannel writer = storage.writer(this, options)) {
+      upload(input, writer);
+    }
+    BlobId blobId = getBlobId();
+    return storage.get(BlobId.of(blobId.getBucket(), blobId.getName()));
+  }
+
+  static void upload(InputStream input, WriteChannel writer) throws IOException {
+    upload(input, writer, DEFAULT_CHUNK_SIZE);
+  }
+
+  /**
+   * Uploads the given content to the storage using specified write channel and the given
+   * buffer size. The default buffer size is 15 MiB. Larger buffer sizes may improve the upload
+   * performance but require more memory. It could cause OutOfMemoryError or add significant garbage
```

Comment:
Do we really need to give the user the option to set the buffer size? Do we provide sufficient tools and documentation to enable someone to do this?

It would be better to simply pick the best buffer size directly. If we're uncertain, lets leave this method out until we are, since it's easier to add it in later than take it out.

Reply:
I think we really need to give the user the option to play with the buffer size. There is no optimal value for the buffer size. By default, the buffer size is equal to the default Write Channel chunk size (15 MiB), but for uploading large files increasing the buffer will give significant performance benefit. Some statistics of uploading 1.5GB file with various buffer sizes:
```
15MiB:  162 seconds
30MiB:  149 seconds
75MiB:  141 seconds
100MiB: 140 seconds
150MiB: 139 seconds
300MiB: 138 seconds
```
On the other hand, one may need to upload hundreds file in parallel, this may require reducing the buffer size to avoid running out of memory. So, giving such flexibility will be certainly a plus.

> Do we provide sufficient tools and documentation to enable someone to do this?

I'm afraid no. There was a request to provide a sort of such documentation: https://github.com/googleapis/java-storage/issues/40 which is closed as duplicate of the current one (add Blob.uploadFrom())

Reply:
In that case this PR needs to include such documentation. I've seen from experience that tunable I/O buffer size parameters are very effective foot-seeking missiles. I'd estimate that somewhere above 80% of invocations of this method will decrease performance.

Reply:
OK, I'll add notice, that this method is for those who want to experiment with the upload time and the amount of the consumed memory.